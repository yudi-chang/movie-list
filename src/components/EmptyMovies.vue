<template>
  <div class="not-found-message">
    <img src="/images/not-found-logo.png" alt="Movie not found" />
    <h2 class="mt-28 text-3xl fw-bold">No movie is found</h2>
  </div>
</template>

<style scoped lang="scss">
.not-found-message {
  text-align: center;
  width: 100%;
  padding: 70px 20px 0 20px;

  img {
    width: 200px;
    max-width: 100%;
    margin: 0 auto;
  }
}
</style>
