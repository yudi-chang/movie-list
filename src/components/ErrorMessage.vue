<template>
  <div class="error-message">
    <img src="/images/error-logo.png" alt="Unexpected error image" />
    <h2 class="mb-12 text-xl fw-bold">Oops something went wrong</h2>
    <div class="message">An unexpected error has occured. please try again or contact support if the error persists</div>
    <button @click="retry()" class="mt-20">
      Try again
    </button>
  </div>
</template>

<script setup lang="ts">
const emit = defineEmits<{
  (event: 'retry-fetch'): void;
}>()

function retry(): void {
  emit('retry-fetch');
}
</script>

<style scoped lang="scss">
.error-message {
  text-align: center;
  width: 100%;
  padding: 100px 20px 0 20px;

  img {
    max-width: 100%;
    margin: 0 auto;
  }
}

.message {
  width: 500px;
  margin: 0 auto;
  max-width: 100%;
  padding: 0 20px;
}

button {
  width: 250px;
  border-radius: 5px;
  background-color: #1289A7;
  color:$secondary-color;
  padding: 10px 20px;
  max-width: 100%;
  cursor: pointer;
  border: none;
  font-size: 17px;
}
</style>
