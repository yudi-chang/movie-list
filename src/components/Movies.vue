
<template>
  <div>
    <Container>
      Movie list
    </Container>
  </div>
</template>

<script setup lang="ts">
import { onBeforeMount } from 'vue';
import type { MoviesFetchPayload } from '@/types';
import { useMovieStore } from '@/stores/movie';
import Container from './Container.vue';

const movieStore = useMovieStore();

onBeforeMount(async () => {
  await movieStore.fetchMovies({} as MoviesFetchPayload);
  console.log("movies: ", movieStore.movies);
});
</script>
