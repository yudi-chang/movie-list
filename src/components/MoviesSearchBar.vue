<template>
  <div class="p-8 search-bar-wrapper">
    <input v-model="titleSearch" @input="debouncedSearch">
    <img src="/images/search-icon.png" alt="Search" />
  </div>
</template>

<script setup lang="ts">
import { ref, inject } from 'vue';
import { useDebounce } from '@/composables/debounce'
import type { MoviesFetchPayload } from '@/types';

// inject
const fetchMovies = inject<(payload: MoviesFetchPayload) => void>('fetchMovies')

// reactive
const titleSearch = ref<string>('')

// methods
const fetchMoviesWithPayload = (): void => {
  if (fetchMovies) {
    const payload: MoviesFetchPayload = {
      Title: titleSearch.value,
      page: 1,
    }
    fetchMovies(payload)
  }
}

// use debounce composable
const { debounce, isWaiting } = useDebounce<string>(() => {
  fetchMoviesWithPayload()
}, 450)

const debouncedSearch = () => {
  debounce()
}
</script>

<style scoped lang="scss">
.search-bar-wrapper {
  background-color: $primary-color;
  position: relative;
  border-radius: 5px;

  input {
    width: 100%;
    height: 50px;
    border-radius: 5px;
    position: relative;
    z-index: 0;
    background-color: #212a31;
    color: $secondary-color;
    transition: border 0.2s ease-in-out;
    padding: 0 45px 0 15px;
  }

  img {
    position: absolute;
    top: 50%;
    right: 25px;
    transform: translateY(-50%);
    height: 20px;
    z-index: 999;
  }
}
</style>