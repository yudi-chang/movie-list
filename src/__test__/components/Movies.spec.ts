import { mount } from '@vue/test-utils';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { createTestingPinia } from '@pinia/testing';
import Movies from '@/components/Movies.vue';
import ErrorMessage from '@/components/ErrorMessage.vue';
import MoviesSearchBar from '@/components/MoviesSearchBar.vue';
import EmptyMovies from '@/components/EmptyMovies.vue';
import FavoritesButton from '@/components/FavoritesButton.vue';
import movieResponse from '@/__test__/mocks/get-movies';

describe('Movie', () => {
  let defaultParams: any;

  const createWrapper = (params: any) => {
    return mount(Movies, params);
  };

  beforeEach(async () => {

    defaultParams = {
      props: {
        movies: [],
        isLoading: false,
        isErrorFetching: false,
        showUtilities: true,
        title: 'Movies',
      },
      global: {
        stubs: {
          ErrorMessage,
          MoviesSearchBar,
          EmptyMovies,
          FavoritesButton
        },
        provide: {
          fetchMovies: vi.fn()
        },
        plugins: [createTestingPinia({
          stubActions: false,
          createSpy: vi.fn,
        })]
      }
    }
  });

  it('should render Movies list correctly when movies is not empty', async () => {
    defaultParams.props.movies = movieResponse.data;
    const wrapper = createWrapper(defaultParams);

    expect(wrapper.find('[data-test="movies-list"]')).toMatchSnapshot();
  });

  it.each([
    { showUtilities: true, description: 'with utilities' },
    { showUtilities: false, description: 'without utilities' }
  ])('should render correctly when component is ($description)', async ({ showUtilities }) => {
    defaultParams.props.movies = movieResponse.data;
    defaultParams.props.showUtilities = showUtilities;
    const wrapper = createWrapper(defaultParams);

    expect(wrapper.find('[data-test="movies-utilities"]').exists()).toBe(showUtilities);
  });

  it('should show loading when loading is true (fetching data)', () => {
    defaultParams.props.isLoading = true;
    const wrapper = createWrapper(defaultParams);

    expect(wrapper.find('[data-test="loader"]').exists()).toBe(true);
    expect(wrapper.findComponent(ErrorMessage).exists()).toBe(false);
    expect(wrapper.find('[data-test="movie-item"]').exists()).toBe(false);
  });

  it('should show error when fetching is error and refetch the data when retry-fetch event is emmited', async () => {
    defaultParams.props.isErrorFetching = true;
    const wrapper = createWrapper(defaultParams);

    expect(wrapper.findComponent(ErrorMessage).exists()).toBe(true);
    expect(wrapper.find('[data-test="loader"]').exists()).toBe(false);
    expect(wrapper.find('[data-test="movie-item"]').exists()).toBe(false);

    await wrapper.findComponent(ErrorMessage).vm.$emit('retry-fetch');
    expect(defaultParams.global.provide.fetchMovies).toHaveBeenCalled();
  });

  it('should show empty state when no movies are available', () => {
    const wrapper = createWrapper(defaultParams);

    expect(wrapper.findComponent(EmptyMovies).exists()).toBe(true);
  });
});