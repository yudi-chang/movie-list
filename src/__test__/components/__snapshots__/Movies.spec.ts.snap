// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Movie > should render Movies list correctly when movies is not empty 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <div
    class="movies mt-40"
    data-test="movies-list"
    data-v-62d96320=""
  >
    
    
    <div
      class="movie-card"
      data-test="movie-item"
      data-v-62d96320=""
    >
      <article
        data-v-62d96320=""
      >
        <div
          class="movie-img"
          data-v-62d96320=""
        >
          <img
            alt="Image of Waterworld"
            data-v-62d96320=""
            loading="lazy"
            src="https://placehold.co/400x500/34495e/FFF?text=Waterworld"
          />
          <div
            class="wrapper fav-btn"
            data-v-13ae506b=""
            data-v-62d96320=""
          >
            <button
              class="star-btn"
              data-test="favorite-btn"
              data-v-13ae506b=""
            >
               ★ 
            </button>
          </div>
        </div>
        <div
          class="movie-info text-md fw-bold p-20 ph-16"
          data-v-62d96320=""
        >
          <h2
            data-v-62d96320=""
          >
            Waterworld
          </h2>
          <div
            class="additional-info"
            data-v-62d96320=""
          >
            <a
              class="imdb-info text-sm ph-12 pv-4 fw-bold"
              data-v-62d96320=""
              href="https://www.imdb.com/title/tt0114898/"
              target="_blank"
            >
               IMDb: tt0114898
            </a>
            <div
              class="text-sm mv-12 ph-12 pv-4 fw-bold year-info"
              data-v-62d96320=""
            >
              Release: 1995
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="movie-card"
      data-test="movie-item"
      data-v-62d96320=""
    >
      <article
        data-v-62d96320=""
      >
        <div
          class="movie-img"
          data-v-62d96320=""
        >
          <img
            alt="Image of Waterworld"
            data-v-62d96320=""
            loading="lazy"
            src="https://placehold.co/400x500/34495e/FFF?text=Waterworld"
          />
          <div
            class="wrapper fav-btn"
            data-v-13ae506b=""
            data-v-62d96320=""
          >
            <button
              class="star-btn"
              data-test="favorite-btn"
              data-v-13ae506b=""
            >
               ★ 
            </button>
          </div>
        </div>
        <div
          class="movie-info text-md fw-bold p-20 ph-16"
          data-v-62d96320=""
        >
          <h2
            data-v-62d96320=""
          >
            Waterworld
          </h2>
          <div
            class="additional-info"
            data-v-62d96320=""
          >
            <a
              class="imdb-info text-sm ph-12 pv-4 fw-bold"
              data-v-62d96320=""
              href="https://www.imdb.com/title/tt0189200/"
              target="_blank"
            >
               IMDb: tt0189200
            </a>
            <div
              class="text-sm mv-12 ph-12 pv-4 fw-bold year-info"
              data-v-62d96320=""
            >
              Release: 1995
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="movie-card"
      data-test="movie-item"
      data-v-62d96320=""
    >
      <article
        data-v-62d96320=""
      >
        <div
          class="movie-img"
          data-v-62d96320=""
        >
          <img
            alt="Image of The Making"
            data-v-62d96320=""
            loading="lazy"
            src="https://placehold.co/400x500/34495e/FFF?text=The+Making"
          />
          <div
            class="wrapper fav-btn"
            data-v-13ae506b=""
            data-v-62d96320=""
          >
            <button
              class="star-btn"
              data-test="favorite-btn"
              data-v-13ae506b=""
            >
               ★ 
            </button>
          </div>
        </div>
        <div
          class="movie-info text-md fw-bold p-20 ph-16"
          data-v-62d96320=""
        >
          <h2
            data-v-62d96320=""
          >
            The Making of 'Waterworld'
          </h2>
          <div
            class="additional-info"
            data-v-62d96320=""
          >
            <a
              class="imdb-info text-sm ph-12 pv-4 fw-bold"
              data-v-62d96320=""
              href="https://www.imdb.com/title/tt2670548/"
              target="_blank"
            >
               IMDb: tt2670548
            </a>
            <div
              class="text-sm mv-12 ph-12 pv-4 fw-bold year-info"
              data-v-62d96320=""
            >
              Release: 1995
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="movie-card"
      data-test="movie-item"
      data-v-62d96320=""
    >
      <article
        data-v-62d96320=""
      >
        <div
          class="movie-img"
          data-v-62d96320=""
        >
          <img
            alt="Image of Waterworld"
            data-v-62d96320=""
            loading="lazy"
            src="https://placehold.co/400x500/34495e/FFF?text=Waterworld"
          />
          <div
            class="wrapper fav-btn"
            data-v-13ae506b=""
            data-v-62d96320=""
          >
            <button
              class="star-btn"
              data-test="favorite-btn"
              data-v-13ae506b=""
            >
               ★ 
            </button>
          </div>
        </div>
        <div
          class="movie-info text-md fw-bold p-20 ph-16"
          data-v-62d96320=""
        >
          <h2
            data-v-62d96320=""
          >
            Waterworld 4: History of the Islands
          </h2>
          <div
            class="additional-info"
            data-v-62d96320=""
          >
            <a
              class="imdb-info text-sm ph-12 pv-4 fw-bold"
              data-v-62d96320=""
              href="https://www.imdb.com/title/tt0161077/"
              target="_blank"
            >
               IMDb: tt0161077
            </a>
            <div
              class="text-sm mv-12 ph-12 pv-4 fw-bold year-info"
              data-v-62d96320=""
            >
              Release: 1997
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="movie-card"
      data-test="movie-item"
      data-v-62d96320=""
    >
      <article
        data-v-62d96320=""
      >
        <div
          class="movie-img"
          data-v-62d96320=""
        >
          <img
            alt="Image of Waterworld"
            data-v-62d96320=""
            loading="lazy"
            src="https://placehold.co/400x500/34495e/FFF?text=Waterworld"
          />
          <div
            class="wrapper fav-btn"
            data-v-13ae506b=""
            data-v-62d96320=""
          >
            <button
              class="star-btn"
              data-test="favorite-btn"
              data-v-13ae506b=""
            >
               ★ 
            </button>
          </div>
        </div>
        <div
          class="movie-info text-md fw-bold p-20 ph-16"
          data-v-62d96320=""
        >
          <h2
            data-v-62d96320=""
          >
            Waterworld
          </h2>
          <div
            class="additional-info"
            data-v-62d96320=""
          >
            <a
              class="imdb-info text-sm ph-12 pv-4 fw-bold"
              data-v-62d96320=""
              href="https://www.imdb.com/title/tt0455840/"
              target="_blank"
            >
               IMDb: tt0455840
            </a>
            <div
              class="text-sm mv-12 ph-12 pv-4 fw-bold year-info"
              data-v-62d96320=""
            >
              Release: 1997
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="movie-card"
      data-test="movie-item"
      data-v-62d96320=""
    >
      <article
        data-v-62d96320=""
      >
        <div
          class="movie-img"
          data-v-62d96320=""
        >
          <img
            alt="Image of Waterworld"
            data-v-62d96320=""
            loading="lazy"
            src="https://placehold.co/400x500/34495e/FFF?text=Waterworld"
          />
          <div
            class="wrapper fav-btn"
            data-v-13ae506b=""
            data-v-62d96320=""
          >
            <button
              class="star-btn"
              data-test="favorite-btn"
              data-v-13ae506b=""
            >
               ★ 
            </button>
          </div>
        </div>
        <div
          class="movie-info text-md fw-bold p-20 ph-16"
          data-v-62d96320=""
        >
          <h2
            data-v-62d96320=""
          >
            Waterworld
          </h2>
          <div
            class="additional-info"
            data-v-62d96320=""
          >
            <a
              class="imdb-info text-sm ph-12 pv-4 fw-bold"
              data-v-62d96320=""
              href="https://www.imdb.com/title/tt0390617/"
              target="_blank"
            >
               IMDb: tt0390617
            </a>
            <div
              class="text-sm mv-12 ph-12 pv-4 fw-bold year-info"
              data-v-62d96320=""
            >
              Release: 1997
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="movie-card"
      data-test="movie-item"
      data-v-62d96320=""
    >
      <article
        data-v-62d96320=""
      >
        <div
          class="movie-img"
          data-v-62d96320=""
        >
          <img
            alt="Image of Swordquest"
            data-v-62d96320=""
            loading="lazy"
            src="https://placehold.co/400x500/34495e/FFF?text=Swordquest"
          />
          <div
            class="wrapper fav-btn"
            data-v-13ae506b=""
            data-v-62d96320=""
          >
            <button
              class="star-btn"
              data-test="favorite-btn"
              data-v-13ae506b=""
            >
               ★ 
            </button>
          </div>
        </div>
        <div
          class="movie-info text-md fw-bold p-20 ph-16"
          data-v-62d96320=""
        >
          <h2
            data-v-62d96320=""
          >
            Swordquest: Waterworld
          </h2>
          <div
            class="additional-info"
            data-v-62d96320=""
          >
            <a
              class="imdb-info text-sm ph-12 pv-4 fw-bold"
              data-v-62d96320=""
              href="https://www.imdb.com/title/tt2761086/"
              target="_blank"
            >
               IMDb: tt2761086
            </a>
            <div
              class="text-sm mv-12 ph-12 pv-4 fw-bold year-info"
              data-v-62d96320=""
            >
              Release: 1983
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="movie-card"
      data-test="movie-item"
      data-v-62d96320=""
    >
      <article
        data-v-62d96320=""
      >
        <div
          class="movie-img"
          data-v-62d96320=""
        >
          <img
            alt="Image of Behind the"
            data-v-62d96320=""
            loading="lazy"
            src="https://placehold.co/400x500/34495e/FFF?text=Behind+the"
          />
          <div
            class="wrapper fav-btn"
            data-v-13ae506b=""
            data-v-62d96320=""
          >
            <button
              class="star-btn"
              data-test="favorite-btn"
              data-v-13ae506b=""
            >
               ★ 
            </button>
          </div>
        </div>
        <div
          class="movie-info text-md fw-bold p-20 ph-16"
          data-v-62d96320=""
        >
          <h2
            data-v-62d96320=""
          >
            Behind the Scenes of the Most Fascinating Waterworld on Earth: The Great Backwaters, Kerala.
          </h2>
          <div
            class="additional-info"
            data-v-62d96320=""
          >
            <a
              class="imdb-info text-sm ph-12 pv-4 fw-bold"
              data-v-62d96320=""
              href="https://www.imdb.com/title/tt5847056/"
              target="_blank"
            >
               IMDb: tt5847056
            </a>
            <div
              class="text-sm mv-12 ph-12 pv-4 fw-bold year-info"
              data-v-62d96320=""
            >
              Release: 2014
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="movie-card"
      data-test="movie-item"
      data-v-62d96320=""
    >
      <article
        data-v-62d96320=""
      >
        <div
          class="movie-img"
          data-v-62d96320=""
        >
          <img
            alt="Image of Louise's W"
            data-v-62d96320=""
            loading="lazy"
            src="https://placehold.co/400x500/34495e/FFF?text=Louise's+W"
          />
          <div
            class="wrapper fav-btn"
            data-v-13ae506b=""
            data-v-62d96320=""
          >
            <button
              class="star-btn"
              data-test="favorite-btn"
              data-v-13ae506b=""
            >
               ★ 
            </button>
          </div>
        </div>
        <div
          class="movie-info text-md fw-bold p-20 ph-16"
          data-v-62d96320=""
        >
          <h2
            data-v-62d96320=""
          >
            Louise's Waterworld
          </h2>
          <div
            class="additional-info"
            data-v-62d96320=""
          >
            <a
              class="imdb-info text-sm ph-12 pv-4 fw-bold"
              data-v-62d96320=""
              href="https://www.imdb.com/title/tt0298417/"
              target="_blank"
            >
               IMDb: tt0298417
            </a>
            <div
              class="text-sm mv-12 ph-12 pv-4 fw-bold year-info"
              data-v-62d96320=""
            >
              Release: 1997
            </div>
          </div>
        </div>
      </article>
    </div>
    <div
      class="movie-card"
      data-test="movie-item"
      data-v-62d96320=""
    >
      <article
        data-v-62d96320=""
      >
        <div
          class="movie-img"
          data-v-62d96320=""
        >
          <img
            alt="Image of Waterworld"
            data-v-62d96320=""
            loading="lazy"
            src="https://placehold.co/400x500/34495e/FFF?text=Waterworld"
          />
          <div
            class="wrapper fav-btn"
            data-v-13ae506b=""
            data-v-62d96320=""
          >
            <button
              class="star-btn"
              data-test="favorite-btn"
              data-v-13ae506b=""
            >
               ★ 
            </button>
          </div>
        </div>
        <div
          class="movie-info text-md fw-bold p-20 ph-16"
          data-v-62d96320=""
        >
          <h2
            data-v-62d96320=""
          >
            Waterworld
          </h2>
          <div
            class="additional-info"
            data-v-62d96320=""
          >
            <a
              class="imdb-info text-sm ph-12 pv-4 fw-bold"
              data-v-62d96320=""
              href="https://www.imdb.com/title/tt0381702/"
              target="_blank"
            >
               IMDb: tt0381702
            </a>
            <div
              class="text-sm mv-12 ph-12 pv-4 fw-bold year-info"
              data-v-62d96320=""
            >
              Release: 2001
            </div>
          </div>
        </div>
      </article>
    </div>
    
    
  </div>,
}
`;
